File: src/main/java/com/example/e_commerce/Service/Produit_service.java
package com.example.e_commerce.Service;

import com.example.e_commerce.Exception.UserNotFoundException;
import com.example.e_commerce.Model.Categorie;
import com.example.e_commerce.Model.Marque;
import com.example.e_commerce.Model.Produit;
import com.example.e_commerce.Model.my_user;
import com.example.e_commerce.Repository.Categorie_Repository;
import com.example.e_commerce.Repository.Marque_Repository;
import com.example.e_commerce.Repository.Produit_Repository;
import com.example.e_commerce.Util.ImageUtils;
import jakarta.transaction.Transactional;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.dao.EmptyResultDataAccessException;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

@Service
@Transactional
@Slf4j
public class Produit_service {

    @Autowired
    private Produit_Repository produitRepository;
    @Autowired
    Marque_Repository marqueRepository;
    @Autowired
    Categorie_Repository categorieRepository;
    public void Add(Produit produit){
        log.info("ADD produit \n");
        produitRepository.save(produit);
    }

    public List<Produit> GetByMarque(List<Long> marqueIds) {
        log.info("Get produits by marqueIds: " + marqueIds);
        try {
            List<Marque> marques = marqueRepository.findAllById(marqueIds);
            if (!marques.isEmpty()) {
                return produitRepository.findByMarqueIn(marques);
            }
        } catch (EmptyResultDataAccessException exception) {
            throw new UserNotFoundException("Marques not found");
        }
        return Collections.emptyList();
    }


    public String uploadImage(MultipartFile file1, String desc_produit, String nom_produit, String prix_produit,
                              String quantitue_produit, String stockage, String garantie, String couleur,
                              String taille_ecran, String ram, String sim, String processeur, String qualite_camera,
                              String graphique,String ref_marque, String ref_categorie) throws IOException {
        Marque marque = marqueRepository.findById(Long.valueOf(ref_marque)).orElse(null);
        Categorie categorie = categorieRepository.findById(Long.valueOf(ref_categorie)).orElse(null);

        if (marque != null && categorie != null) {
            Produit produit = Produit.builder()
                    .name1(file1.getOriginalFilename())
                    .type1(file1.getContentType())
                    .imageData1(ImageUtils.compressImage(file1.getBytes()))
                    .desc_produit(desc_produit)
                    .nom_produit(nom_produit)
                    .prix_produit(prix_produit)
                    .quantitue_produit(quantitue_produit)
                    .stockage(stockage)
                    .garantie(garantie)
                    .couleur(couleur)
                    .taille_ecran(taille_ecran)
                    .ram(ram)
                    .sim(sim)
                    .processeur(processeur)
                    .qualite_camera(qualite_camera)
                    .graphique(graphique)
                    .marque(marque)
                    .categorie(categorie)
                    .build();

            Produit savedProduit = produitRepository.save(produit);
            if (savedProduit != null) {
                return "file uploaded successfully: image 1: " + file1.getOriginalFilename();
            }
        }

        return null;
    }


    public byte[] downloadImage(Long ref_produit){
        Optional<Produit> dbImageData = produitRepository.findById(ref_produit);
        byte[] images= ImageUtils.decompressImage(dbImageData.get().getImageData1());
        return images;
    }

    public List<Produit> GetAll(){
        log.info("Get all produit");
        return  produitRepository.findAll();
    }

    public Produit GetOne(Long ref_produit){
        log.info("Get one with ref_produit "+ref_produit);
        try {
            return produitRepository.findById(ref_produit).get();
        }catch (EmptyResultDataAccessException exception){
            throw new UserNotFoundException("produit with ref_produit "+ref_produit+" not found");
        }
    }
    public Produit GetOnemarque(Long ref_marque){
        log.info("Get one with ref marque "+ref_marque);
        try {
            return produitRepository.findById(ref_marque).get();
        }catch (EmptyResultDataAccessException exception){
            throw new UserNotFoundException("produit with ref_produit "+ref_marque+" not found");
        }
    }



    public void Delete(Long ref_produit){
        log.info("Delete produit "+ref_produit);
        try {
            produitRepository.deleteById(ref_produit);
        }catch (EmptyResultDataAccessException exception){
            throw new UserNotFoundException("produit with ref_produit "+ref_produit+" not found");
        }
    }
    public void Update_quantite(Long ref_produit,String quantite){
        log.info("update produit quantité");
        try {
            Produit produit = GetOne(ref_produit);
            produit.setQuantitue_produit(quantite);
            Add(produit);
        }catch (EmptyResultDataAccessException exception){
            throw new UserNotFoundException("produit with ref "+ref_produit+" not found");
        }
    }
    public void quantite(Long ref_produit,int quantite){
        log.info("update produit quantité");
        try {
            Produit produit = GetOne(ref_produit);
            produit.setQuantite(quantite);
            Add(produit);
        }catch (EmptyResultDataAccessException exception){
            throw new UserNotFoundException("produit with ref "+ref_produit+" not found");
        }
    }
    public void Update_description(Long ref_produit,String description){
        log.info("update produit description");
        try {
            Produit produit = GetOne(ref_produit);
            produit.setDesc_produit(description);
            Add(produit);
        }catch (EmptyResultDataAccessException exception){
            throw new UserNotFoundException("produit with ref "+ref_produit+" not found");
        }
    }
    public void Update_nomproduit(Long ref_produit,String nom_produit){
        log.info("update produit quantité");
        try {
            Produit produit = GetOne(ref_produit);
            produit.setNom_produit(nom_produit);
            Add(produit);
        }catch (EmptyResultDataAccessException exception){
            throw new UserNotFoundException("produit with ref "+ref_produit+" not found");
        }
    }    public void Update_prix(Long ref_produit,String prix){
        log.info("update produit prix");
        try {
            Produit produit = GetOne(ref_produit);
            produit.setPrix_produit(prix);
            Add(produit);
        }catch (EmptyResultDataAccessException exception){
            throw new UserNotFoundException("produit with ref "+ref_produit+" not found");
        }
    }
    public void update_all_fileds(Produit produit){
        log.info("update_all_fileds");
        produitRepository.save(produit);
    }

    public void Update_stockage(Long ref_caracteristique,String stockage){
        log.info("update Stockage");
        try {
            Produit caracteristique = GetOne(ref_caracteristique);
            caracteristique.setStockage(stockage);
            Add(caracteristique);
        }catch (EmptyResultDataAccessException exception){
            throw new UserNotFoundException("Caracteristique with ref_caracteristique "+ref_caracteristique+" not found");
        }
    }
    public void Update_garantie(Long ref_caracteristique,String garantie){
        log.info("update garantie");
        try {
            Produit caracteristique = GetOne(ref_caracteristique);
            caracteristique.setGarantie(garantie);
            Add(caracteristique);
        }catch (EmptyResultDataAccessException exception){
            throw new UserNotFoundException("Caracteristique with ref_caracteristique "+ref_caracteristique+" not found");
        }
    }
    public void Update_couleur(Long ref_caracteristique,String couleur){
        log.info("update couleur");
        try {
            Produit caracteristique = GetOne(ref_caracteristique);
            caracteristique.setCouleur(couleur);
            Add(caracteristique);
        }catch (EmptyResultDataAccessException exception){
            throw new UserNotFoundException("Caracteristique with ref_caracteristique "+ref_caracteristique+" not found");
        }
    }
    public void Update_taille_ecran(Long ref_caracteristique,String taille_ecran){
        log.info("update taille_ecran");
        try {
            Produit caracteristique = GetOne(ref_caracteristique);
            caracteristique.setTaille_ecran(taille_ecran);
            Add(caracteristique);
        }catch (EmptyResultDataAccessException exception){
            throw new UserNotFoundException("Caracteristique with ref_caracteristique "+ref_caracteristique+" not found");
        }
    }
    public void Update_ram(Long ref_caracteristique,String ram){
        log.info("update ram");
        try {
            Produit caracteristique = GetOne(ref_caracteristique);
            caracteristique.setRam(ram);
            Add(caracteristique);
        }catch (EmptyResultDataAccessException exception){
            throw new UserNotFoundException("Caracteristique with ref_caracteristique "+ref_caracteristique+" not found");
        }
    }
    public void Update_sim(Long ref_caracteristique,String sim){
        log.info("update sim");
        try {
            Produit caracteristique = GetOne(ref_caracteristique);
            caracteristique.setRam(sim);
            Add(caracteristique);
        }catch (EmptyResultDataAccessException exception){
            throw new UserNotFoundException("Caracteristique with ref_caracteristique "+ref_caracteristique+" not found");
        }
    }
    public void Update_processeur(Long ref_caracteristique,String processeur){
        log.info("update processeur");
        try {
            Produit caracteristique = GetOne(ref_caracteristique);
            caracteristique.setProcesseur(processeur);
            Add(caracteristique);
        }catch (EmptyResultDataAccessException exception){
            throw new UserNotFoundException("Caracteristique with ref_caracteristique "+ref_caracteristique+" not found");
        }
    }
    public Produit getProductById(Long id) {

        return produitRepository.findById(id).get();
    }

}

File: frontend/src/app/services/produit.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

import { FormBuilder, FormGroup, FormControl, ReactiveFormsModule, Validators }from '@angular/forms';
import { produit } from '../model/produit';
@Injectable({
  providedIn: 'root'
})
export class ProduitService {

  public dataForm: FormGroup;
  private apiUrl = 'http://localhost:8080'; // Remplacez par l'URL de votre API


  constructor(private http :  HttpClient) { }
  public GetAll():Observable<any>
  {
    return this.http.get('http://localhost:8080/api/produit/GetAll');
  }

  public GetOne(id : number)
  {
    return this.http.get('http://localhost:8080/api/produit/GetOne/'+id)
  }

  public GetOnemarque(ref_marque: string | number) {
    return this.http.get<any>('http://localhost:8080/api/produit/GetOnemarque/' + ref_marque);
  }

  GetByMarque(marqueIds: number[]): Observable<produit[]> {
    const params = new HttpParams().set('marqueIds', marqueIds.join(','));
    const url = `${this.apiUrl}/api/produit/GetByMarque`;
    return this.http.get<produit[]>(url, { params });
  }
  public GetProduitsByCategorie(ref_categorie: number): Observable<any> {
    const url = `http://localhost:8080/api/categorie/GetProduitsByCategorie/${ref_categorie}`;
    return this.http.get(url);
  }

  public update(obj : any):Observable<any>
  {
    return this.http.put('http://localhost:8080/api/produit/update_all',obj,{responseType : 'text'})
  }






  public Delete(id : Number)
  {
    return this.http.delete('http://localhost:8080/api/produit/Delete/'+id,{responseType : 'text'})
  }

  public Register(formData : FormData):Observable<any> {
    return this.http.post('http://localhost:8080/api/produit/Add', formData, {responseType : 'text'});
  }
  uploadImage(file1: File, desc_produit: string, nom_produit: string, prix_produit: string,
    quantitue_produit: string, stockage: string, garantie: string, couleur: string,
    taille_ecran: string, ram: string, sim: string, processeur: string, qualite_camera: string,
    graphique: string, ref_marque: string, ref_categorie: string): Observable<any> {
const formData = new FormData();
formData.append('image1', file1);
formData.append('desc_produit', desc_produit);
formData.append('nom_produit', nom_produit);
formData.append('prix_produit', prix_produit);
formData.append('quantitue_produit', quantitue_produit);
formData.append('stockage', stockage);
formData.append('garantie', garantie);
formData.append('couleur', couleur);
formData.append('taille_ecran', taille_ecran);
formData.append('ram', ram);
formData.append('sim', sim);
formData.append('processeur', processeur);
formData.append('qualite_camera', qualite_camera);
formData.append('graphique', graphique);
formData.append('ref_marque', (ref_marque));
formData.append('ref_categorie', (ref_categorie));

return this.http.post('http://localhost:8080/api/produit/addimage', formData, { responseType: 'text' });
}


  // public Register(formData: FormData): Observable<any> {
  //   return this.http.post('http://localhost:8080/api/marque/Add', formData, { responseType: 'text' });
  // }

  public savemarque(formData: FormData): Observable<any> {
    return this.http.post('http://localhost:8080/api/produit/savemarque', formData, { responseType: 'text' });
  }
}
